tuf/repository_lib.py

```diff
@@ -1549,6 +1566,222 @@ def _get_hashes_and_length_if_needed(use_length, use_hashes, full_file_path,
+  # This last node will be the root of the tree.
+  current_nodes = leaves
+
+  while(len(current_nodes) > 1):
+    new_nodes = []
+    for i in range(0, len(current_nodes), 2):
+      # If there are an odd number of nodes and this is the last
+      # node, add this node to the next level.
+      if i + 1 >= len(current_nodes):
+        new_nodes.append(current_nodes[i])
+      # Otherwise, use the next two nodes to build a new node.
+      else:
+        n = InternalNode(current_nodes[i], current_nodes[i+1])
+        # Add this node to the next level
+        new_nodes.append(n)
+    current_nodes = new_nodes
+
+  # The only node remaining in current_nodes will be the root node.
+  root = current_nodes[0]
+
+  # Return the root node and the leaves. The root hash must be used along with the
+  # path to verify the tree. The root hash should be securely sent to
+  # each client. To do so, we will add it to the timestamp metadata.
+  # The leaves will be used to find the path to each leaf and send
+  # this path to the client for verification
+  return root, leaves
+
+def _write_merkle_paths(root, leaves, storage_backend, merkle_directory, version):
+  # The root and leaves must be part of the same fully constructed
+  # Merkle tree. Create a path from
+  # Each leaf to the root node. This path will be downloaded by
+  # the client and used for verification of the tree. For each
+  # step in the path, keep track of both the sibling node and
+  # Whether this is a left or a right child.
+
+  # Before writing each leaf, make sure the storage_backend
+  # is instantiated
+  if storage_backend is None:
+    storage_backend = securesystemslib.storage.FilesystemBackend()
+
+  for l in leaves:
+    merkle_path = {}
+    current_node = l
+    path_directions = {}
+
+    index = 0
+
+    while(current_node != root):
+      next_node = current_node.parent
+      # TODO: determine left or right upon node creation.
+      # This currently determines which sibling to use by
+      # finding the sibling that does not match the current digest.
+      h_left = next_node.left.digest
+      h_right = next_node.right.digest
+      if current_node.digest == h_left:
+        merkle_path[str(index)] = h_right
+        path_directions[str(index)] = -1
+      elif current_node.digest== h_right:
+        merkle_path[str(index)] = h_left
+        path_directions[str(index)] = 1
+      else:
+        # error
+        pass
+      index = index + 1
+
+      current_node = next_node
+
+    # Write the path to the merkle_directory
+    file_contents = tuf.formats.build_dict_conforming_to_schema(
+        tuf.formats.SNAPSHOT_MERKLE_SCHEMA,
+        leaf_contents=l.contents,
+        merkle_path=merkle_path,
+        path_directions=path_directions)
+    file_content = _get_written_metadata(file_contents)
+    file_object = tempfile.TemporaryFile()
+    file_object.write(file_content)
+    filename = os.path.join(merkle_directory, l.name + '-snapshot.json')
```

Comment:
How does this directory scale relative to the sum of the leaf nodes? Would it still be required if we included a key field with every child pointer so a search through the tree would be used rather than a directory?

Reply:
This directory is for storing all of the nodes on disc (which might look different for a non-filesystem storage mechanism). The key field might help with discovery of the nodes, but they'd still all need to be stored somewhere.